src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,256 @@
+            break;
+          }
+        }
+
+        // Since we stop when we find the first line with a less indentation
+        // we pull back one line to get to the end of the function/class
+        idx--;
+
+        const endLine = lines[idx];
+
+        structure.push({
+          type,
+          start: {
+            line: info.line,
+            character: info.indentation,
+          },
+          end: {
+            line: endLine.line,
+            character: endLine.text.search(/(?<=\S)\s*$/) - 1, // Calculate position of last non-white character
+          },
+        });
+      }
+    }
+
+    return structure;
+  }
+
+  /*
+   * Filter function that returns true if an element is strictly ahead (false if equal)
+   * of the specified (first arg) position.
+   */
+  static isAhead(position: MinimalPosition, elementPosition: MinimalPosition) {
+    return (
+      elementPosition.line > position.line ||
+      (elementPosition.line === position.line && elementPosition.character > position.character)
+    );
+  }
+
+  /*
+   * Filter function that returns true if an element is strictly behind (false if equal)
+   * of the specified (first arg) position.
+   */
+  static isBehind(position: MinimalPosition, elementPosition: MinimalPosition) {
+    return (
+      elementPosition.line < position.line ||
+      (elementPosition.line === position.line && elementPosition.character < position.character)
+    );
+  }
+
+  /*
+   * Find the position of the specified:
+   *    type: function or class
+   *    direction: next or prev
+   *    edge: start or end
+   *
+   * With this information one can determine all of the required motions
+   */
+  find(type: Type, direction: Direction, edge: Edge, position: Position): Position | null {
+    // Choose the filtering utility based on direction
+    const isDirection = direction === 'next' ? PythonDocument.isAhead : PythonDocument.isBehind;
+
+    // Filter function for all elements whose "edge" is in the correct "direction"
+    // relative to the cursor's position
+    const dir = (element: StructureElement) => isDirection(position, element[edge]);
+
+    // Filter out elements from structure based on type and direction
+    const elements = this.structure.filter((elem) => elem.type === type).filter(dir);
+
+    if (edge === 'end') {
+      // When moving to an 'end' the elements should be started by the end position
+      elements.sort((a, b) => a.end.line - b.end.line);
+    }
+
+    // Return the first match if any exist
+    if (elements.length) {
+      // If direction === 'next' return the first element
+      // otherwise return the last element
+      const index = direction === 'next' ? 0 : elements.length - 1;
+      const element = elements[index];
+      const pos = element[edge];
+
+      // execAction MUST return a fully realized Position object created using new
+      return new Position(pos.line, pos.character);
+    }
+
+    return null;
```

Comment:
Use `undefined` instead of `null`

Reply:
The logic at the call-site uses `null` to denote an alternate (failed) return state. I considered `undefined` but I thought `null` represents this idea better since `undefined` is more implicit.

I am open to change this.

Reply:
Yeah, since the type system in typescript makes the usual use of `undefined` in javascript (checking whether a property exists on an object) useless, they're generally equivalent.
For that reason, I like to keep it simple and pretend `null` doesn't exist (the typescript team [does the same](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined), though it isn't an official recommendation, just their preference).

Reply:
Done